src/generators/utils/parse.ts

```diff
@@ -47,3 +55,132 @@ export function parseSitesFixesConfig<T extends SiteProps>(text: string, options
 
     return sites;
 }
+
+// URL patterns are guaranteed to not have protocol and leading '/'
+function getDomain(url: string) {
+    return url.split('/')[0].toLowerCase();
+}
+
+/*
+ * Encode all offsets into a string, where each record is 6 bytes long:
+ *  - 4 bytes for start offset
+ *  - 3 bytes for record length (end offset - start offset)
+ * Both values are stored in base 36 (radix 36) notation.
+ * Maximum supported numbers:
+ *  - start offset must be no more than parseInt('zzzz', 36) = 1679615
+ *  - length must be no more than parseInt('zzz', 36) = 46655
+ *
+ * We have to encode offsets into a string to be able to save them in
+ * chrome.storage.local for use in non-persistent background contexts.
+ */
+function encodeOffsets(offsets: Array<[number, number]>): string {
+    return offsets.map(([offset, length]) => {
+        const stringOffset = offset.toString(36);
+        const stringLength = length.toString(36);
+        return '0'.repeat(4 - stringOffset.length) + stringOffset + '0'.repeat(3 - stringLength.length) + stringLength;
+    }).join('');
+}
+
+function decodeOffset(offsets: string, index: number): [number, number] {
+    const base = (4 + 3) * index;
+    const offset = parseInt(offsets.substring(base + 0, base + 4), 36);
+    const length = parseInt(offsets.substring(base + 4, base + 4 + 3), 36);
+    return [
+        offset,
+        offset + length,
+    ];
+}
+
+export function indexSitesFixesConfig<T extends SiteProps>(text: string): SitePropsIndex<T> {
+    const domains: {[domain: string]: number | number[]} = {};
+    const domainPatterns: {[domainPattern: string]: number | number[]} = {};
+    // Array of tuples, where first number is an offset of record start and second number is record length.
+    const offsets: Array<[number, number]> = [];
+
+    function processBlock(recordStart: number, recordEnd: number, index: number) {
+        // TODO: more formal definition of URLs and delimiters
+        const block = text.substring(recordStart, recordEnd);
+        const lines = block.split('\n');
+        const commandIndices: number[] = [];
+        lines.forEach((ln, i) => {
+            if (ln.match(/^[A-Z]+(\s[A-Z]+){0,2}$/)) {
```

Comment:
Not sure if there will be no commands with 3 and more words in future.

Reply:
This restriction was suggested by @Gusted, I'll defer to him on this.

Reply:
I live in `new Date()` not in `new Date(new Date().setFullYear(new Date().getFullYear() + 1)) ` ;)

Reply:
Let's leave this restriction as is and then relax it later if ever need to.

Reply:
I can live with that.